File: commonfamily/src/main/java/com/koenig/commonModel/finance/Balance.java
package com.koenig.commonModel.finance;

import com.koenig.commonModel.Item;

import org.joda.time.DateTime;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Thomas on 23.12.2017.
 */

public class Balance extends Item {
    private int balance;
    private DateTime date;

    public Balance(String id, int balance, DateTime dateTime) {
        super(id, Integer.toString(balance));
        this.balance = balance;
        this.date = dateTime;
    }

    public Balance(int balance, DateTime dateTime) {
        super(Integer.toString(balance));
        this.balance = balance;
        this.date = dateTime;
    }

    public Balance(ByteBuffer buffer) {
        super(buffer);
        balance = buffer.getInt();
        date = Companion.byteToDateTime(buffer);
    }

    public static List<Balance> getBalances(byte[] bytes) {
        ByteBuffer buffer = ByteBuffer.wrap(bytes);
        int size = buffer.getInt();
        ArrayList<Balance> balances = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            balances.add(new Balance(buffer));
        }

        return balances;
    }

    public static byte[] listToBytes(List<Balance> balances) {
        int size = 4;
        for (Balance balance : balances) {
            size += balance.getByteLength();
        }

        ByteBuffer buffer = ByteBuffer.allocate(size);
        buffer.putInt(balances.size());
        for (Balance balance : balances) {
            balance.writeBytes(buffer);
        }
        return buffer.array();
    }
    public int getBalance() {
        return balance;
    }

    public void setBalance(int balance) {
        this.balance = balance;
    }

    public DateTime getDate() {
        return date;
    }

    public void setDate(DateTime date) {
        this.date = date;
    }

    @Override
    public int getByteLength() {
        return super.getByteLength() + 4 + Companion.getDateLength();
    }

    @Override
    public void writeBytes(ByteBuffer buffer) {
        super.writeBytes(buffer);
        buffer.putInt(balance);
        Companion.writeDateTime(date, buffer);
    }

    @Override
    public String toString() {
        return "Balance{" +
                "balance=" + balance +
                ", date=" + date +
                '}';
    }
}

File: commonfamily/src/main/java/com/koenig/commonModel/finance/Balance.kt
package com.koenig.commonModel.finance

import com.koenig.commonModel.Item
import com.koenig.commonModel.byteLength
import com.koenig.commonModel.localDate
import com.koenig.commonModel.writeBytes
import org.joda.time.LocalDate
import java.nio.ByteBuffer
import java.util.*

/**
 * Created by Thomas on 23.12.2017.
 */

class Balance : Item {
    var balance: Int = 0
    var day: LocalDate

    override val byteLength: Int
        get() = super.byteLength + 4 + day.byteLength

    constructor(id: String, balance: Int, day: LocalDate) : super(id, Integer.toString(balance)) {
        this.balance = balance
        this.day = day
    }

    constructor(balance: Int, day: LocalDate) : super(Integer.toString(balance)) {
        this.balance = balance
        this.day = day
    }

    constructor(buffer: ByteBuffer) : super(buffer) {
        balance = buffer.int
        day = buffer.localDate
    }

    override fun writeBytes(buffer: ByteBuffer) {
        super.writeBytes(buffer)
        buffer.putInt(balance)
        day.writeBytes(buffer)
    }

    override fun toString(): String {
        return "Balance{" +
                "balance=" + balance +
                ", day=" + day +
                '}'.toString()
    }

    companion object {

        fun getBalances(bytes: ByteArray): MutableList<Balance> {
            val buffer = ByteBuffer.wrap(bytes)
            val size = buffer.int
            val balances = ArrayList<Balance>(size)
            for (i in 0 until size) {
                balances.add(Balance(buffer))
            }

            return balances
        }

        fun listToBytes(balances: List<Balance>): ByteArray {
            val size = 4 + balances.sumBy { it.byteLength }

            val buffer = ByteBuffer.allocate(size)
            buffer.putInt(balances.size)
            for (balance in balances) {
                balance.writeBytes(buffer)
            }
            return buffer.array()
        }
    }
}
